game/groundforces/template_loader.py

```diff
@@ -0,0 +1,361 @@
+        and dumps them to a pickle"""
+        mappings: dict[str, list[TemplateMapping]] = {}
+        with logged_duration("Parsing mapping yamls"):
+            for file in Path(TEMPLATE_DIR).rglob("*.yaml"):
+                if not file.is_file():
+                    continue
+                with file.open("r", encoding="utf-8") as f:
+                    mapping_dict = yaml.safe_load(f)
+
+                template_map = TemplateMapping.from_dict(mapping_dict, f.name)
+
+                if template_map.template_file in mappings:
+                    mappings[template_map.template_file].append(template_map)
+                else:
+                    mappings[template_map.template_file] = [template_map]
+
+        with logged_duration(f"Loading templates multithreaded"):
+            with ThreadPoolExecutor(max_workers=4) as exe:
+                for miz, maps in mappings.items():
+                    exe.submit(self._load_from_miz, miz, maps)
+
+        self._dump_templates()
+
+    def _dump_templates(self) -> None:
+        file = Path(persistency.base_path()) / TEMPLATE_DUMP
+        dump = (VERSION, self._templates)
+        with file.open("wb") as fdata:
+            pickle.dump(dump, fdata)
+
+    @staticmethod
+    def mapping_for_group(
+        mappings: list[TemplateMapping], group_name: str
+    ) -> tuple[TemplateMapping, int, GroupTemplateMapping]:
+        for mapping in mappings:
+            for g_id, group_mapping in enumerate(mapping.groups):
+                if (
+                    group_mapping.name == group_name
+                    or group_name in group_mapping.statics
+                ):
+                    return mapping, g_id, group_mapping
+        raise KeyError
+
+    def _load_from_miz(self, miz: str, mappings: list[TemplateMapping]) -> None:
+        template_position: dict[str, Point] = {}
+        temp_mis = dcs.Mission()
+        with logged_duration("dcs mission loading"):
+            # The load_file takes a lot of time to compute. That's why the templates
+            # are written to a pickle and can be reloaded from the ui
+            # Example the whole routine: 0:00:00.934417,
+            # the .load_file() method: 0:00:00.920409
+            temp_mis.load_file(miz)
+
+        for country in itertools.chain(
+            temp_mis.coalition["red"].countries.values(),
+            temp_mis.coalition["blue"].countries.values(),
+        ):
+            for dcs_group in itertools.chain(
+                temp_mis.country(country.name).vehicle_group,
+                temp_mis.country(country.name).ship_group,
+                temp_mis.country(country.name).static_group,
+            ):
+                try:
+                    mapping, group_id, group_mapping = self.mapping_for_group(
+                        mappings, dcs_group.name
+                    )
+                except KeyError:
+                    logging.error(f"No mapping for dcs group {dcs_group.name}")
```

Comment:
I'm pretty sure I've asked this before... by why isn't this a stronger error? Worthy of a comment

Reply:
I even degraded it as a warning. This KeyError means that we do not have a mapping for a unit/group placed in one of the layout miz. For us this will not break anything. On the other hand the devloper who messed with the layout should know that something went wrong and his/her layout will not work properly. But i do personally not feel good about raising an exception for such a mistake within a layout.